src/storage.rs

```diff
@@ -96,166 +112,237 @@ pub trait Storage {
 /// The Memory Storage Core instance holds the actual state of the storage struct. To access this
 /// value, use the `rl` and `wl` functions on the main MemStorage implementation.
 pub struct MemStorageCore {
-    hard_state: HardState,
-    snapshot: Snapshot,
-    // TODO: maybe vec_deque
+    raft_state: RaftState,
     // entries[i] has raft log position i+snapshot.get_metadata().get_index()
     entries: Vec<Entry>,
+    // Metadata of the last snapshot received.
+    snapshot_metadata: SnapshotMetadata,
 }
 
 impl Default for MemStorageCore {
     fn default() -> MemStorageCore {
         MemStorageCore {
+            raft_state: Default::default(),
             // When starting from scratch populate the list with a dummy entry at term zero.
+            // The dummy entry is used to ensure `entries` never be empty, which can simplify
+            // many logics.
             entries: vec![Entry::new()],
-            hard_state: HardState::new(),
-            snapshot: Snapshot::new(),
+            // Every time a snapshot is applied to the storage, the metadata will be stored here,
+            // and a dummy entry will be pushed to `entries`.
+            snapshot_metadata: Default::default(),
         }
     }
 }
 
 impl MemStorageCore {
+    /// Append a configuration state. Only for tests.
+    pub fn append_conf_state(&mut self, cs: ConfStateWithIndex) {
```

Comment:
Seems unrelated to this PR?

Reply:
It's used in `test_membership_change.rs`, for persisting configuration before power cycle.